sandbox/controller.go

```diff
@@ -0,0 +1,63 @@
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package sandbox
+
+import (
+	"context"
+
+	"github.com/containerd/containerd/api/services/sandbox/v1"
+)
+
+// Controller is an interface to manage sandboxes at runtime.
+// When running the traditional containerd shim, the workflow looks as follows:
+// For each new task we're about to run:
+//  1. Invoke `shim_binary --start` to obtain `TaskService` address (printed in stdout)
+//  2. Call TaskService.RunContainer(id=1)
+//  3. Exec `shim_binary --delete` to stop shim
+//  4. Exec `shim_binary --start` again to obtain another `TaskService` address
+//  5. TaskService.RunContainer(id=2)
+//  6. Exec `shim_binary --delete` to stop shim
+//
+// When running in sandbox mode, shim must implement `SandboxService`.
+// In sandbox mode shim lifetimes are managed manually via sandbox API.
+//  1. Client calls `client.SandboxController.Start()` to launch new shim and create sandbox process
+//  2. Run containers with `shim.TaskService.RunContainer(id=1)` and another one `shim.TaskService.RunContainer(id=2)`
+//  3. ... usual container lifecycle calls to `shim.TaskService`
+//  4. Client calls shim to stop the sandbox with `client.SandboxService.Shutdown()`
+//  5. Shim implementation will perform cleanup similar to regular task service (e.g. shutdown, clean, and `shim_binary --delete`)
+type Controller interface {
+	// Start will start new sandbox instance.
+	// containerd will run new shim runtime instance and will invoke Start to create a sandbox process.
+	// This routine must be invoked before scheduling containers on this instance.
+	// Once started clients may run containers via Task service (additionally specifying sandbox id the container will belong to).
+	Start(ctx context.Context, sandboxID string) (uint32, error)
+	// Shutdown deletes and cleans all tasks and sandbox instance.
+	Shutdown(ctx context.Context, sandboxID string) error
+	// Wait blocks until sandbox process exits.
+	Wait(ctx context.Context, sandboxID string) (*sandbox.ControllerWaitResponse, error)
+	// Pause will freeze running sandbox instance.
+	// Shim implementations may return ErrNotImplemented if this is out of scope of a given sandbox.
+	Pause(ctx context.Context, sandboxID string) error
+	// Resume will unfreeze previously paused sandbox instance
+	Resume(ctx context.Context, sandboxID string) error
+	// Ping is a lightweight API call to check whether sandbox instance is still alive (e.g. quick livability check).
+	// This should not involve any complex logic and containerd will not debug log it as it might be called quite often.
+	Ping(ctx context.Context, sandboxID string) error
```

Comment:
How is this intended to be used? I'm wondering if it is really necessary in most cases vs using `Status`. If/when it is necessary, should we focus on ways of making the API extensible. The same extensibility argument could be made for `Pause`/`Resume` as well

Reply:
`Ping` expected to be used to check shim's livability (e.g. whether shim process / whatever behind responds to client's requests), typically this can be useful when restoring shims on the daemon restart (kin of `TaskService.Connect`).

`Status` aims to provide basic metadata about sandbox state (pid, state, exit code, status, etc).

As for `Pause/Resume` I recall we had such API in firecracker-containerd to control microVM process: [link](https://github.com/firecracker-microvm/firecracker-containerd/blob/321527e24556ad51279b7b7bb4c65b59c518b5d3/firecracker-control/local.go#L252)
@kzys may have more thoughts whether we need it / want it. I'm fine to postpone it to later stages.

Reply:
If Status covers the common use case, fewer is better. For Pause/Resume, I'm not questioning if they are useful, just whether they should be part of the core interface rather than implemented as a separate plugin. I could imagine more functionality going along with it. If clients already have to deal with unimplemented, splitting them into separate services makes sense.

Reply:
Addressed in https://github.com/containerd/containerd/pull/6703/commits/1125871820e146aecfa91912021b7fd65cd6fdac